addons/binding/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/handler/EnergenieGatewayHandler.java

```diff
@@ -0,0 +1,619 @@
+            if (!isGatewayThingAlreadyRegistered(thing)) {
+                logger.debug("About to register a new gateway");
+                registerNewGateway();
+            } else {
+                logger.warn(ALREADY_REGISTERED_GATEWAY_MESSAGE);
+                updateStatus(ThingStatus.REMOVED);
+            }
+        } else {
+            gatewayId = Integer.parseInt(thing.getProperties().get(EnergenieBindingConstants.PROPERTY_DEVICE_ID));
+        }
+        updateStatus(ThingStatus.ONLINE);
+        scheduleUpdateTask(this.updateInterval);
+    }
+
+    private boolean isGatewayComingFromDiscoveryService(Thing gatewayThing) {
+        Map<String, String> props = gatewayThing.getProperties();
+        String authCode = props.get(EnergenieBindingConstants.PROPERTY_AUTH_CODE);
+        return authCode != null;
+    }
+
+    /**
+     * Checks if there is already a registered thing for a given gateway device
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private boolean isGatewayThingAlreadyRegistered(Thing thing) {
+        ServiceReference ref = bundleContext.getServiceReference(ThingRegistry.class.getName());
+        thingRegistry = (ThingRegistry) bundleContext.getService(ref);
+        Collection<Thing> things = thingRegistry.getAll();
+        for (Thing registeredThing : things) {
+            ThingUID registeredThingUID = registeredThing.getUID();
+            ThingTypeUID registeredThingTypeUID = registeredThing.getThingTypeUID();
+            if (registeredThingTypeUID != null
+                    && registeredThingTypeUID.equals(EnergenieBindingConstants.THING_TYPE_GATEWAY)) {
+                Configuration config = registeredThing.getConfiguration();
+                String code = (String) config.get(EnergenieBindingConstants.CONFIG_GATEWAY_CODE);
+                if (gatewayCode != null && gatewayCode.equals(code) && (!thing.getUID().equals(registeredThingUID))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public void registerNewGateway() {
+        JsonObject registrationResponse = apiManager.registerGateway(getThing().getLabel(), gatewayCode);
+        if (registrationResponse != null) {
+            JsonObject responseData = registrationResponse.get(JSONResponseConstants.DATA_KEY).getAsJsonObject();
+            gatewayId = responseData.get(DeviceConstants.DEVICE_ID_KEY).getAsInt();
+            setProperties(responseData);
+        } else {
+            logger.warn(NOT_SUCCESSFULL_REGISTRATION_MESSAGE);
+        }
+    }
+
+    private void scheduleUpdateTask(long updateInterval) {
+        if (updateTaskResult != null) {
+            this.updateTaskResult.cancel(true);
+            this.updateTaskResult = null;
+        }
+        /**
+         * The refresh thread which is used for checking the "last_seen_at"
+         * property of the gateway
+         */
+        Runnable refreshRunnable = () -> {
```

Comment:
I think the idea of the Java 8 style is to inline this, but I might be wrong :-)